Mathlib/RingTheory/TwoSidedIdeal/Operations.lean

```diff
@@ -0,0 +1,199 @@
+homomorphism.
+-/
+def map (I : TwoSidedIdeal R) : TwoSidedIdeal S :=
+  span (f '' I)
+
+lemma map_mono {I J : TwoSidedIdeal R} (h : I ≤ J) :
+    map f I ≤ map f J :=
+  span_mono $ Set.image_mono h
+
+/--
+kernel of a ring homomorphism as a two-sided-ideal.
+-/
+def ker (f : R →+* S) : TwoSidedIdeal R :=
+  .mk'
+    {r | f r = 0} (map_zero _) (by rintro _ _ (h1 : f _ = 0) (h2 : f _ = 0); simp [h1, h2])
+    (by rintro _ (h : f _ = 0); simp [h]) (by rintro _ _ (h : f _ = 0); simp [h])
+    (by rintro _ _ (h : f _ = 0); simp [h])
+
+lemma mem_ker (f : R →+* S) {x : R} : x ∈ ker f ↔ f x = 0 := by
+  delta ker; rw [mem_mk']; rfl
+  · rintro _ _ (h1 : f _ = 0) (h2 : f _ = 0); simp [h1, h2]
+  · rintro _ (h : f _ = 0); simp [h]
+  · rintro _ _ (h : f _ = 0); simp [h]
+  · rintro _ _ (h : f _ = 0); simp [h]
+
+end NonAssocRing
+
+section Ring
+
+variable {R : Type*} [Ring R]
+
+lemma mem_span_iff_exists {s : Set R} {x} :
+    x ∈ span s ↔
+    ∃ (ι : Type) (fin : Fintype ι) (xL : ι → R) (xR : ι → R) (y : ι → s),
+      x = ∑ i : ι, xL i * (y i : R) * xR i := by
+  let S : TwoSidedIdeal R := .mk'
+    {x | ∃ (ι : Type) (fin : Fintype ι) (xL : ι → R) (xR : ι → R) (y : ι → s),
+    x = ∑ i : ι, xL i * y i * xR i}
+    ⟨Empty, Fintype.instEmpty, Empty.elim, Empty.elim, Empty.elim, by simp⟩
+    (by
+      rintro _ _ ⟨na, fina, xLa, xRa, ya, rfl⟩ ⟨nb, finb, xLb, xRb, yb, rfl⟩
+      refine ⟨na ⊕ nb, inferInstance, Sum.elim xLa xLb, Sum.elim xRa xRb, Sum.elim ya yb, by simp⟩)
+    (by
+      rintro _ ⟨n, finn, xL, xR, y, rfl⟩
+      exact ⟨n, finn, -xL, xR, y, by simp⟩)
+    (by
+      rintro a _ ⟨n, finn, xL, xR, y, rfl⟩
+      exact ⟨n, finn, a • xL, xR, y, by simp [Finset.mul_sum, mul_assoc]⟩)
+    (by
+      rintro _ b ⟨n, finn, xL, xR, y, rfl⟩
+      exact ⟨n, finn, xL, fun x ↦ xR x * b, y, by simp [Finset.sum_mul, mul_assoc]⟩)
+  suffices eq1 : span s = S by
+    rw [eq1]; simp [S, mem_mk']
+  rw [span, RingCon.ringConGen_eq]
+  refine ringCon_injective $ sInf_eq_of_forall_ge_of_forall_gt_exists_lt ?_ ?_
+  · rintro I (hI : ∀ a b, _ → _)
+    rw [← ringCon_le_iff, le_iff]
+    intro x h
+    simp only [SetLike.mem_coe, mem_mk', Set.mem_setOf_eq, S] at h
+    obtain ⟨n, finn, xL, xR, y, rfl⟩ := h
+    refine finsetSum_mem _ _ _ fun i _ => mul_mem_right _ _ _ $
+      mul_mem_left _ _ _ $  hI (y i) 0 (by simp)
+  · rintro I hI
+    exact ⟨S.ringCon, fun a b H ↦ ⟨PUnit, inferInstance, fun _ ↦ 1, fun _ ↦ 1,
+      fun _ ↦ ⟨a - b, by simp [H]⟩, by simp⟩, hI⟩
+
+/-- Every two-sided-ideal is also a left-ideal. -/
+def asIdeal : TwoSidedIdeal R →o Ideal R where
+  toFun I :=
+  { carrier := I
+    add_mem' := I.add_mem
+    zero_mem' := I.zero_mem
+    smul_mem' := fun r x hx => I.mul_mem_left r x hx }
+  monotone' _ _ h _ h' := h h'
+
+/--Given an ideal `I`, `span I` is the smallest two-sided-ideal containing `I`.-/
+def fromIdeal : Ideal R →o TwoSidedIdeal R where
```

Comment:
I think this, and below, either need `@[simps]` or hand written simp lemmas.

Reply:
I have handwritten `mem_...` lemmas, `simps` lemma will be about `ringCon` which hopeful would be avoided after enough API